johnnyfive/confluence.py

```diff
@@ -0,0 +1,274 @@
+        for this method is to be used in concert with `get_page_contents` to
+        obtain a page, modify it, then replace it.
+
+        Parameters
+        ----------
+        body : `str`
+            The new page contents to upload to Confluence.
+
+        Returns
+        -------
+        _type_
+            _description_
+        """
+        safe_confluence_connect(self.instance.update_page, self.page_id,
+                                self.title, body)
+
+    def _set_metadata(self):
+        """_set_metadata Set the various instance metadata
+
+        Especially after a page is created or deleted, this method updates the
+        various instance attributes to keep current.
+        """
+        self.exists = safe_confluence_connect(self.instance.page_exists,
+                                              self.space, self.title)
+
+        # Page-Specific Information
+        self.page_id = None if not self.exists else \
+            safe_confluence_connect(self.instance.get_page_id,
+                                    self.space, self.title)
+        self.attachment_url = None if not self.exists else \
+            f"{self.instance.url}download/attachments/{self.page_id}/"
+
+
+# Internal Functions =========================================================#
+def setup_confluence():
+    """setup_confluence Set up the Confluence class instance
+
+    Reads in the confluence.conf configuration file, which contains the URL,
+    username, and password.  Also contained in the configuration file are
+    the Confluence space and page title into which the updated table will be
+    placed.
+
+    Returns
+    -------
+    confluence : `atlassian.Confluence`
+        Confluence class, initialized with credentials
+    """
+    # Read the setup
+    setup = utils.read_ligmos_conffiles('confluenceSetup')
+
+    # Return the Confluence instance
+    return Confluence( url=setup.host,
+                       username=setup.user,
+                       password=setup.password )
+
+
+def safe_confluence_connect(func, *args, pause=5, timeout=5, **kwargs):
+    """safe_confluence_connect Safely connect to Confluence (error-catching)
+
+    Wrapper for confluence-connection functions to catch errors that might be
+    kicked (ConnectionTimeout, for instance).
+
+    This function performs a semi-infinite loop, pausing for 5 seconds after
+    each failed function call, up to a maximum of 5 minutes.
+
+    Parameters
+    ----------
+    func : `method`
+        The Confluence class method to be wrapped
+    pause : `int` or `float`, optional
+        The number of seconds to wait in between retries to connect.
+        [Default: 5]
+    timeout : `int` or `float`, optional
+        The total time (in minutes) to continue retrying before returning None
+        [Default: 5]
+
+    Returns
+    -------
+    `Any`
+        The return value of `func` -- or None if unable to run `func`
+    """
+    # Starting value
+    i = 1
+
+    while True:
```

Comment:
I'd like to avoid this if possible; I see the escape hatch a little further down, but it'd be safer to just try to connect N times and raise an exception if the connection fails, or set the instance to None or something so it's easier to catch.  The site internet/network config isn't robust enough and this will likely get hung up and have to be manually killed/restarted on occasion.

Reply:
Converted the `while...break` syntax into a `for...` loop.